File: KnowitJulekalender2023/Dag01/Dag01.fs
open System.IO
open System.Text.RegularExpressions

type Bet = { GoalsBetted : int; Odds : float; }

let parseBet (str : string) =
    let parts = str.Split(", ")
    { GoalsBetted = int parts[0]; Odds = double parts[1];}

let evaluateBet goals bet =
    match goals with
        | _unused when goals >= bet.GoalsBetted -> bet.Odds
        | _ -> 0

let applyBet total winFactor =
    let betfactor = 0.175
    let amountBet = round (betfactor * total)
    match winFactor with 
        | 0.0 -> total - amountBet
        | _ -> total + round (amountBet * winFactor)
    
let bets = Regex.Matches(File.ReadAllText("bets.txt"), "\[(.*?)\]")
            |> Seq.map (fun m -> m.Groups[1].Value)
            |> Seq.map parseBet

let goals = File.ReadAllText("goals.txt").Split(",") 
            |> Seq.map int

let initial = 50_000

Seq.map2 evaluateBet goals bets
|> Seq.fold applyBet initial
|> (fun left -> initial - int(left))
|> printfn "Loss: %A" 

File: KnowitJulekalender2023/Dag01_CS/Dag01.cs
using System.Text.RegularExpressions;

var bets = Regex.Matches(File.ReadAllText("bets.txt"), @"\[(\d), (\d\.\d+)\]")
    .Select(m => new { 
        GoalsBet = int.Parse(m.Groups[1].Value), 
        Odds = double.Parse(m.Groups[2].Value) 
    })
    .ToList();

var goalsScored = File.ReadAllText("goals.txt")
    .Split(",")
    .Select(s => int.Parse(s))
    .ToList();

var outcomes = bets.Zip(goalsScored);

double cash = 50_000;
double betFactor = 0.175f;

foreach(var outcome in outcomes)
{
    var bet = outcome.First;
    var result = outcome.Second;

    var amountBet = Math.Round(cash * betFactor);
    var winFactor = bet.GoalsBet > result ? 0.0f : bet.Odds;
    if (winFactor > 0) cash += Math.Round(amountBet * winFactor);
    else cash -= amountBet;
}

Console.WriteLine("Gjenværende stenger: " + cash);
Console.WriteLine("Antall stenger tapt: " + (50_000 - cash));
